src/lib/structures/Argument.ts

```diff
@@ -77,17 +83,25 @@ export interface IArgument<T> {
  *   }
  *
  *   run(argument) {
+ *     const resolved = CoreArgument.resolve(argument);
+ *     if (resolved.success) return this.ok(resolved.value);
+ *     return this.error(argument, 'ArgumentHyperlinkInvalidURL', resolved.error);
+ *   }
+ *
+ *   static resolve(argument) {
  *     try {
- *       return this.ok(new URL(argument));
+ *       return new URL(ok(argument));
  *     } catch {
- *       return this.error(argument, 'ArgumentHyperlinkInvalidURL', 'The argument did not resolve to a valid URL.');
+ *       err('The argument did not resolve to a valid URL.');
  *     }
  *   }
  * }
  * ```
  */
 export abstract class Argument<T = unknown> extends AliasPiece implements IArgument<T> {
 	public abstract run(parameter: string, context: ArgumentContext<T>): ArgumentResult<T>;
+	// TODO: Add definition of 'public static abstract resolve' here whenever typescript supports it.
```

Comment:
Ah you added this. Good call there. For what it's worth, now we can keep track of it.

Reply:
And that's... when?

Reply:
Maybe never, who knows, I don't see how that's a problem though. It's just a reminder, just like the many TODOs we have in Skyra.

Reply:
To be honest, I'm not a fan of this PR, the arguments are already accessible *everywhere*, even outside client, why do we need this, and what stops them from using `container.stores.get('arguments').get('bool') as Argument<boolean>`?

If argument augmentation is needed, we already do that by offering the ExtendedArgument, something that already exists.

Furthermore, this PR completely kills any attempt we may have at library agnosticism, as some arguments are discord.js specific.

Reply:
> and what stops them from using container.stores.get('arguments').get('bool') as Argument<boolean>?

Well it requires a command doesn't it?

> Furthermore, this PR completely kills any attempt we may have at library agnosticism, as some arguments are discord.js specific.

I can probably use container values as suggested by Favna below, rather than parameters.

Reply:
> Well it requires a command doesn't it?

I'll make `Args#run` a thing to abstract it, so as long as you have an `Args` instance, you'll be able to dynamically run any argument:

```typescript
const argument = this.container.stores.get('arguments').get('integer') as Argument<number>;
const times = await args.run('5', argument);
// times -> 5
```

I could also augment ArgumentStore like I did with [StoreRegistry](https://github.com/sapphiredev/framework/blob/af30ab09580cb44917388ef1dde9bacc3fb5118e/src/lib/structures/StoreRegistry.ts#L84-L100), so the cast above is unnecessary.

> I can probably use container values as suggested by Favna below, rather than parameters.

I have no idea what this means, please elaborate.

Reply:
Replied in main thread